app/services/coupons/preview_service.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+module Coupons
+  class PreviewService < BaseService
+    def initialize(invoice:, applied_coupons:)
+      @invoice = invoice
+      @applied_coupons = applied_coupons
+
+      super
+    end
+
+    def call
+      return result.not_found_failure!(resource: 'invoice') unless invoice
+      return result.not_found_failure!(resource: 'applied_coupons') unless applied_coupons
+
+      result.credits = []
+
+      applied_coupons.each do |applied_coupon|
+        break unless invoice.sub_total_excluding_taxes_amount_cents&.positive?
+        next unless invoice.currency == applied_coupon.amount_currency
+
+        fees = fees(applied_coupon)
+
+        next if fees.none?
+
+        base_amount_cents = base_amount_cents(applied_coupon, fees)
+        credit = add_credit(applied_coupon, fees, base_amount_cents)
+
+        result.credits << credit
+        invoice.credits << credit
+      end
+
+      result.invoice = invoice
+      result
+    end
+
+    private
+
+    attr_reader :applied_coupons, :invoice
+
+    def add_credit(applied_coupon, fees, base_amount_cents)
+      credit_amount = AppliedCoupons::AmountService.call(applied_coupon:, base_amount_cents:).amount
+      new_credit = Credit.new(
+        invoice:,
+        applied_coupon:,
+        amount_cents: credit_amount,
+        amount_currency: invoice.currency,
+        before_taxes: true
+      )
```

Comment:
when is the credit saved?

Reply:
For preview invoice we compute everything on the fly. That is the main difference between this service and service that is called on real invoice. There are also several other differences (frequency or remaining amount is not tracked...)

Reply:
then it is expected to not persist the credits instantiated here.

Reply:
exactly, this is expected not to store credit